sparse/coo/indexing.py

```diff
@@ -0,0 +1,450 @@
+
+            if start != stop:
+                starts.append(start)
+                stops.append(stop)
+                n_matches += stop - start
+
+    return starts, stops, n_matches
+
+
+@numba.jit(nopython=True)
+def _get_slice_len(idx):  # pragma: no cover
+    """
+    Get the number of elements in a slice.
+
+    Parameters
+    ----------
+    idx : np.ndarray
+        A (3,) shaped array containing start, stop, step
+
+    Returns
+    -------
+    n : int
+        The length of the slice.
+
+    Examples
+    --------
+    >>> idx = np.array([5, 15, 5])
+    >>> _get_slice_len(idx)
+    2
+    """
+    start, stop, step = idx[0], idx[1], idx[2]
+
+    if step > 0:
+        return (stop - start + step - 1) // step
+    else:
+        return (start - stop - step - 1) // (-step)
+
+
+@numba.jit(nopython=True)
+def _filter_pairs(starts, stops, coords, indices):  # pragma: no cover
+    """
+    Converts all the pairs into a single integer mask, additionally filtering
+    by the indices.
+
+    Parameters
+    ----------
+    starts, stops : list[int]
+        The starts and stops to convert into an array.
+    coords : np.ndarray
+        The coordinates to filter by.
+    indices : np.ndarray
+        The indices in the form of slices such that indices[:, 0] are starts,
+        indices[:, 1] are stops and indices[:, 2] are steps.
+
+    Returns
+    -------
+    mask : list
+        The output integer mask.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> starts = [2]
+    >>> stops = [7]
+    >>> coords = np.array([[0, 1, 2, 3, 4, 5, 6, 7]])
+    >>> indices = np.array([[2, 8, 2]]) # Start, stop, step pairs
+    >>> _filter_pairs(starts, stops, coords, indices)
+    [2, 4, 6]
+    """
+    mask = []
+
+    # For each pair,
+    for i in range(len(starts)):
+        # For each element match within the pair range
+        for j in range(starts[i], stops[i]):
+            match = True
+
+            # Check if it matches all indices
+            for k in range(len(indices)):
+                idx = indices[k]
+                elem = coords[k, j]
+
+                match &= ((elem - idx[0]) % idx[2] == 0 and
+                          ((idx[2] > 0 and idx[0] <= elem < idx[1])
+                           or (idx[2] < 0 and idx[0] >= elem > idx[1])))
+
+            # and append to the mask if so.
+            if match:
+                mask.append(j)
+
+    return mask
+
+
+@numba.jit(nopython=True)
+def _join_adjacent_pairs(starts_old, stops_old):  # pragma: no cover
+    """
+    Joins adjacent pairs into one. For example, 2-5 and 5-7
+    will reduce to 2-7 (a single pair). This may help in
+    returning a slice in the end which could be faster.
+
+    Parameters
+    ----------
+    starts_old, stops_old : list[int]
+        The input starts and stops
+    Returns
+    -------
+    starts, stops : list[int]
+        The reduced starts and stops.
+    Examples
```

Comment:
Style nit, add line breaks before section titles.

Reply:
I'm not too used to RST, I believe that's required for clean rendering as well.